contracts/lambdas/Dex_lambdas.ligo

```diff
@@ -8,120 +8,74 @@ function initialize_exchange(
-          }
-        else skip;
+      const result: tmp_tokens_type = Set.fold(
+        get_tokens_from_param,
+        params.input_tokens,
+        default_tmp_tokens);
+
+      const tokens : tokens_type = result.tokens;
+      const token_bytes : bytes = Bytes.pack(tokens);
+      var (pair_i, token_id) := get_pair_info(token_bytes,
+        s.pools_count,
+        s.pool_to_id,
+        s.pools);
+      s.tokens[token_id] := tokens;
 
-      const (pair_i, token_id) = get_pair_info(tokens, s);
+      assert_with_error(pair_i.total_supply = 0n, ERRORS.pair_listed);
 
       if s.pools_count = token_id
       then {
-        s.pool_to_id[Bytes.pack(tokens)] := token_id;
+        s.pool_to_id[token_bytes] := token_id;
         s.pools_count := s.pools_count + 1n;
       }
       else skip;
 
-      s.tokens[token_id] := tokens;
-
-      function map_rates_outs_zeros(
-        const acc   : (
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat)
-        );
-        const entry : (token_pool_index * input_token)
-      )             : (
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat) *
-          map(token_pool_index, nat)
-        ) is (
-          Map.add(entry.0, entry.1.rate,                      acc.0),
-          Map.add(entry.0, entry.1.precision_multiplier,      acc.1),
-          Map.add(entry.0, entry.1.in_amount,                 acc.2),
-          Map.add(entry.0, 0n,                                acc.3)
-        );
-
-      const (token_rates, precision_multipliers, inputs, zeros) = Map.fold(
-        map_rates_outs_zeros,
-        params.input_tokens,
-        (
-          (map[]: map(token_pool_index, nat)),
-          (map[]: map(token_pool_index, nat)),
-          (map[]: map(token_pool_index, nat)),
-          (map[]: map(token_pool_index, nat))
-        )
-      );
-
-      if pair_i.total_supply =/= 0n
-      then failwith(ERRORS.pair_listed)
-      else skip;
-
-      var new_pair: pair_type := pair_i;
-      if CONSTANTS.max_a < params.a_constant
-      then failwith("A const limit")
-      else skip;
-      new_pair.initial_A := params.a_constant;
-      new_pair.future_A := params.a_constant;
-      new_pair.initial_A_time := Tezos.now;
-      new_pair.future_A_time := Tezos.now;
-      new_pair.token_rates := token_rates;
-      new_pair.precision_multipliers := precision_multipliers;
-      new_pair.reserves := zeros;
-      new_pair.virtual_reserves := zeros;
-      new_pair.proxy_limits := zeros;
+      assert_with_error(
+        CONSTANTS.max_a >= params.a_constant,
+        ERRORS.a_limit);
+
+      patch pair_i with record [
+        initial_A = params.a_constant;
+        future_A = params.a_constant;
+        initial_A_time = Tezos.now;
+        future_A_time = Tezos.now;
+        tokens_info = params.tokens_info;
```

Comment:
If we set reserves here in the first investment would be doubled reserves. Reserves should change in `add_liq` function